### PR TITLE
add name attribute to Linter class

### LINKED ISSUE #746: linter exclusions bug
Running `lint_package(linters = object_name_linter())` on current master of lintr:

```
R/utils.R:224:1: style: Variable and function name style should be snake_case or symbols.
Linter <- function(fun) { # nolint: object_name_linter.
^~~~~~
```

and a couple of similar issues.

Note that if the `nolint` comment for `Linter` in `R/utils.R` is changed, removing the terminal dot:
```
# R/utils.R
Linter <- function(fun) { # nolint: object_name_linter
  # blah
}
```

... no lint is thrown for the name `Linter`

### PR DESCRIPTION
**Pull request by AshesITR**
fixes #746

Things that should be done in this PR:

 - [x] remove `linter` argument from `Lint`. It's overwritten during `lint()` anyway.
 - [x] refactor `expect_lint()` to nevertheless recognize `linter` as a `lint_fields` despite not being an argument of `Lint()`.
 - [x] document the new behaviour.
 - [x] add NEWS bullet.

### CODE REVIEW

File: R/function_left_parentheses.R [hunk 1]
```diff
@@ -7,7 +7,7 @@ function_left_parentheses_linter <- function() { # nolint: object_length_linter.
       ids_with_token(source_file, "'('"),
       function(id) {
 
-        parsed <- source_file$parsed_content[id, ]
+        parsed <- source_file$parsed_content[id,]
```
Comment (MichaelChirico) [inline R/function_left_parentheses.R:10]: the lint is correct, isn't it?

File: R/utils.R [hunk 2]
```diff
@@ -62,13 +62,30 @@ names2 <- function(x) {
   names(x) %||% rep("", length(x))
 }
 
+linter_auto_name <- function(which = -3L) {
+  call <- sys.call(which = which)
+  nm <- paste(deparse(call, 500L), collapse = "")
```
Comment (MichaelChirico) [inline R/utils.R:67]: `collapse` should be `" "` IINM

File: R/utils.R [hunk 3]
```diff
@@ -62,13 +62,30 @@ names2 <- function(x) {
...
   missing <- nms == ""
   if (all(!missing)) return(nms)
 
-  deparse2 <- function(x) paste(deparse(x, 500L), collapse = "")
-  defaults <- vapply(x[missing], deparse2, character(1), USE.NAMES = FALSE)
+  default_name <- function(x) {
+    if (inherits(x, "linter")) {
+      attr(x, "name", exact = TRUE)
+    } else {
+      paste(deparse(x, 500L), collapse = "")
```
Comment (MichaelChirico) [inline R/utils.R:85]: ditto

File: tests/testthat/test-methods.R [hunk 4]
```diff
@@ -22,27 +22,34 @@ test_that("it returns the input trimmed to the last full lint if one exists with
...
+      type = "error",
+      message = "Under no circumstances is the use of foobar allowed.",
+      line = "a <- 1",
+      ranges = list(c(1, 2), c(10, 20))),
     "lint"
   )
 
+  expect_warning(
+    Lint("dummy.R", linter = "deprecated"),
+    fixed = "deprecated"
```
Comment (MichaelChirico) [inline tests/testthat/test-methods.R:50]: isn't `fixed` the argument passed to `grepl`?

File: tests/testthat/test-methods.R [hunk 5]
```diff
@@ -57,8 +64,9 @@ test_that("as.data.frame.lints", {
     type = c("style", "error"),
     message = c("", "Under no circumstances is the use of foobar allowed."),
     line = c("", "a <- 1"),
-    linter = c("", "custom_linter"),
-    stringsAsFactors = FALSE)
+    linter = c(NA_character_, NA_character_), # These are assigned in lint() now.
```
Comment (MichaelChirico) [inline tests/testthat/test-methods.R:67]: shouldn't `as.data.frame.lints` just exclude this column now? Or keeping it for now for backwards compatibility?

**Review by MichaelChirico — CHANGES_REQUESTED** [review]
Basically looks good, small changes

File: tests/testthat/test-methods.R (hunk 5 above)
Comment (AshesITR) [reply tests/testthat/test-methods.R:67]: It's still very useful, as it doesn't contain NAs when the `lints` come from an actual call to `lint()`.

File: tests/testthat/test-methods.R (hunk 4 above)
Comment (AshesITR) [reply tests/testthat/test-methods.R:50]: Yep, the warning should contain the text "deprecated", no?

Comment (MichaelChirico) [reply tests/testthat/test-methods.R:50]: then fixed=TRUE and I think warning="deprecated" if you want to name it right?

Comment (AshesITR) [reply tests/testthat/test-methods.R:50]: Oh, thanks for pointing that out.
We have a lot of these calls in `test-lint_file.R`

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**